File: backend/app/api/data/wealth.py
"""재테크 로드맵 (목표금액 → 달성계획 + 자격조건별 상품추천) endpoints."""
from __future__ import annotations

from fastapi import APIRouter, Body, Depends, Query

from app.core.auth import require_auth

from app.data.market import wealthplan
from app.data.market import picks

router = APIRouter()


@router.get("/wealth/plan")
def wealth_plan(user: str = Depends(require_auth)):
    """목표 달성 계획 + 자격 상품 추천 + 로드맵."""
    return wealthplan.get_plan(user)


@router.post("/wealth/profile")
def wealth_profile(profile: dict = Body(...), user: str = Depends(require_auth)):
    """프로필/목표 저장(나이·연봉·월수입·월저축·현재자산·결혼·무주택·자녀·목표금액·기간) → 계획 반환."""
    return wealthplan.save_profile(user, profile)


@router.get("/wealth/loan-sim")
def wealth_loan_sim(loan_amount: float = Query(...), loan_rate: float = Query(default=6.0),
                    loan_years: int = Query(default=5, ge=1, le=40),
                    invest_return: float = Query(default=8.0),
                    user: str = Depends(require_auth)):
    """대출 레버리지 시뮬 — 월 상환액·총이자·투자 예상가치·순손익·손익분기 수익률·위험 경고."""
    return wealthplan.loan_sim(loan_amount, loan_rate, loan_years, invest_return)


@router.get("/wealth/realty-sim")
def wealth_realty_sim(price: float = Query(...), own_capital: float = Query(...),
                      loan_rate: float = Query(default=4.5), years: int = Query(default=5, ge=1, le=40),
                      appreciation: float = Query(default=3.0),
                      mode: str = Query(default="wolse"),
                      deposit: float = Query(default=0.0), rent_monthly: float = Query(default=0.0),
                      user: str = Depends(require_auth)):
    """부동산 투자 시뮬 — 자기자본+대출로 매입해 전세/월세. 대출·월현금흐름·수익률(ROE)·집값 시나리오·위험."""
    return wealthplan.realty_sim(price, own_capital, loan_rate, years, appreciation, mode, deposit, rent_monthly)


@router.get("/wealth/holdings")
def wealth_holdings(user: str = Depends(require_auth)):
    """내가 하고 있는 저축·상품 + 혜택 + N년 뒤 예상 금액(정부매칭·세제혜택 포함)."""
    return wealthplan.get_holdings(user)


@router.post("/wealth/holdings")
def wealth_holdings_save(body: dict = Body(...), user: str = Depends(require_auth)):
    """내 저축·상품 저장 → 예상 재계산. body: {holdings:[{name,monthly,current}], horizon:int}"""
    return wealthplan.save_holdings(user, body.get("holdings", []), body.get("horizon", 10))


@router.get("/wealth/dividend-sim")
def wealth_dividend_sim(invest: float = Query(default=100000000), yield_pct: float = Query(default=5.0),
                        years: int = Query(default=10, ge=1, le=40), growth_pct: float = Query(default=3.0),
                        reinvest: bool = Query(default=True), user: str = Depends(require_auth)):
    """배당주 소득 — 투자금→연/월 배당(세후)·재투자 경로·목표 월소득에 필요한 투자금 + 방법 가이드."""
    return wealthplan.dividend_plan(invest, yield_pct, years, growth_pct, reinvest)


@router.get("/wealth/ipo-sim")
def wealth_ipo_sim(offer_price: float = Query(default=30000), alloc_shares: float = Query(default=10),
                   subscribe_amount: float = Query(default=0), user: str = Depends(require_auth)):
    """공모주(IPO) 소득 — 배정 주수·공모가→상장일 상승률별 수익 시나리오 + 청약 방법 가이드."""
    return wealthplan.ipo_plan(offer_price, alloc_shares, subscribe_amount)


@router.get("/wealth/realty-loans")
def wealth_realty_loans(price: float = Query(...), annual_income: float = Query(default=0),
                        age: float = Query(default=0), married: bool = Query(default=False),
                        homeless: bool = Query(default=True), has_child: bool = Query(default=False),
                        deposit: float = Query(default=0), mode: str = Query(default="wolse"),
                        user: str = Depends(require_auth)):
    """부동산 대출 종류·한도 — 매매가·소득·자격으로 주담대(LTV·DSR)·디딤돌·보금자리·신생아특례·전세자금 한도."""
    return wealthplan.realty_loans(price, annual_income, age, married, homeless, has_child, deposit, mode)


@router.get("/wealth/dividend-picks")
def wealth_dividend_picks(top: int = Query(default=12, ge=1, le=40), user: str = Depends(require_auth)):
    """배당주 추천 — 고배당 상위(유동성 필터·2~15% 밴드) + 1천만 투자 시 세후 월배당. 실시간 스냅샷."""
    return picks.dividend_picks(top)


@router.get("/wealth/ipo-schedule")
def wealth_ipo_schedule(user: str = Depends(require_auth)):
    """공모주 청약일정 — 38커뮤니케이션 스크래핑(캐시 30분). 청약중·예정·최근마감 종목·공모가밴드·주간사."""
    return picks.ipo_schedule()

File: frontend/src/lib/api/wealth.ts
// 재테크 로드맵 — 목표계획·자격 상품추천·대출/부동산/배당/공모주 시뮬

import { request } from "./client";

export interface WealthProduct {
  name: string;
  category: string;
  eligible: boolean;
  cond: string;
  benefit: string;
  cap: number;
  priority: number;
  link: string;
  example: string;
}

export interface WealthPlan {
  profile: Record<string, unknown>;
  goal: { amount: number; years: number };
  required_monthly: number;
  capacity_monthly: number;
  feasible: boolean | null;
  shortfall: number;
  reach_years: number | null;
  assumed_return: number;
  scenarios: {
    key: string; name: string; desc: string; safety: string; risk: string;
    return_mid: number; return_low: number; return_high: number;
    balance_at_goal_years: number; balance_low: number; balance_high: number;
    reach_years: number | null; reach_years_low: number | null; reach_years_high: number | null;
    time_saved_vs_safe: number | null; recommended: boolean;
  }[];
  projection: { year: number; balance: number }[];
  products: WealthProduct[];
  eligible_count: number;
  allocation: { name: string; monthly: number; category: string; why: string }[];
  steps: string[];
  note: string;
}

export interface LoanSim {
  loan_amount: number; loan_rate: number; loan_years: number; invest_return: number;
  monthly_payment: number; total_repay: number; total_interest: number;
  invest_value: number; net_profit: number; worthwhile: boolean;
  breakeven_return: number | null;
  scenarios: { name: string; return: number; invest_value: number; net_profit: number; worthwhile: boolean }[];
  loans: { name: string; rate: number; note: string }[];
  verdict: string; warning: string;
}

export interface RealtySim {
  mode: string; price: number; own_capital: number; loan: number; loan_rate: number; years: number; appreciation: number;
  deposit: number; rent_monthly: number; monthly_interest: number; monthly_cashflow: number;
  rent_yield_on_capital: number | null; total_interest: number;
  future_price: number; sale_gain: number; net_profit: number; roe: number | null; roe_no_leverage: number;
  scenarios: { name: string; appreciation: number; future_price: number; sale_gain: number; net_profit: number; roe: number | null }[];
  note: string; warning: string;
}

export interface RealtyLoan {
  name: string; kind: string; rate: number; limit: number | null; eligible: boolean; cond: string; note: string;
}

export interface RealtyLoans {
  price: number; annual_income: number; mode: string; ltv_pct: number;
  loans: RealtyLoan[]; eligible_count: number; max_limit: number; dsr_note: string; note: string;
}

export interface HoldingCatalogItem {
  name: string; category: string; benefit: string; example: string; rate: number; bonus_note: string; has_bonus: boolean;
}

export interface HoldingItem {
  name: string; category: string; monthly: number; current: number; rate: number; bonus_note: string;
  principal: number; invest_value: number; bonus_total: number; total: number; gain: number;
  yearly: { year: number; total: number }[];
}

export interface HoldingsData {
  holdings: { name: string; monthly: number; current: number }[];
  horizon: number;
  catalog: HoldingCatalogItem[];
  projection: {
    horizon: number;
    items: HoldingItem[];
    totals_by_year: { year: number; total: number }[];
    summary: { monthly_sum: number; principal: number; bonus_total: number; gain: number; total: number };
    note: string;
  };
}

export interface DividendSim {
  invest: number; yield_pct: number; years: number; growth_pct: number; reinvest: boolean; tax_pct: number;
  annual_gross: number; annual_net: number; monthly_net: number; final_value: number; total_dividends_net: number;
  yearly: { year: number; dividend_net: number; cum_net: number; value: number }[];
  targets: { monthly: number; invest: number }[];
  examples: { name: string; yield: string; note: string }[];
  guide: string[]; note: string;
}

export interface IpoSim {
  offer_price: number; alloc_shares: number; cost: number; subscribe_amount: number; margin_estimate: number;
  scenarios: { gain_pct: number; sell_price: number; profit: number; roi_on_cost: number | null }[];
  guide: string[]; note: string;
}

export interface DividendPick {
  ticker: string; name: string; sector: string | null; close: number | null;
  div_yield: number; per: number | null; pbr: number | null; roe: number | null;
  market_cap: number | null; foreign_ratio: number | null; op_yoy: number | null;
  score: number; grade: string; reasons: string[]; stability: string; cycle: string;
  monthly_per_10m: number; naver_url: string;
}

export interface DividendPicks { generated_at: string; picks: DividendPick[]; guide: string[]; note: string; }

export interface IpoScheduleItem {
  no?: string; name: string; subscribe: string; status: string; price_confirmed: string | null; price_band: string; underwriter: string;
  market?: string; shares?: string; offer_amount_text?: string; offer_amount_won?: number | null;
  listing_date?: string; demand_competition?: string; lockup?: string; detail_url?: string;
}

export interface IpoSchedule {
  items: IpoScheduleItem[]; upcoming_count: number; source: string; generated_at: string; error?: string; guide?: string[]; note: string;
}

export const wealthApi = {
  wealthPlan: () => request<WealthPlan>(`/api/data/wealth/plan`),
  wealthSaveProfile: (profile: Record<string, unknown>) =>
    request<WealthPlan>(`/api/data/wealth/profile`, { method: "POST", body: JSON.stringify(profile) }),
  wealthLoanSim: (loanAmount: number, loanRate: number, loanYears: number, investReturn: number) =>
    request<LoanSim>(`/api/data/wealth/loan-sim?loan_amount=${loanAmount}&loan_rate=${loanRate}&loan_years=${loanYears}&invest_return=${investReturn}`),
  wealthRealtySim: (p: { price: number; own_capital: number; loan_rate: number; years: number; appreciation: number; mode: string; deposit: number; rent_monthly: number }) =>
    request<RealtySim>(`/api/data/wealth/realty-sim?price=${p.price}&own_capital=${p.own_capital}&loan_rate=${p.loan_rate}&years=${p.years}&appreciation=${p.appreciation}&mode=${p.mode}&deposit=${p.deposit}&rent_monthly=${p.rent_monthly}`),
  wealthRealtyLoans: (p: { price: number; annual_income: number; age: number; married: boolean; homeless: boolean; has_child: boolean; deposit: number; mode: string }) =>
    request<RealtyLoans>(`/api/data/wealth/realty-loans?price=${p.price}&annual_income=${p.annual_income}&age=${p.age}&married=${p.married}&homeless=${p.homeless}&has_child=${p.has_child}&deposit=${p.deposit}&mode=${p.mode}`),
  wealthHoldings: () => request<HoldingsData>(`/api/data/wealth/holdings`),
  wealthSaveHoldings: (holdings: { name: string; monthly: number; current: number }[], horizon: number) =>
    request<HoldingsData>(`/api/data/wealth/holdings`, { method: "POST", body: JSON.stringify({ holdings, horizon }) }),
  wealthDividendSim: (p: { invest: number; yield_pct: number; years: number; growth_pct: number; reinvest: boolean }) =>
    request<DividendSim>(`/api/data/wealth/dividend-sim?invest=${p.invest}&yield_pct=${p.yield_pct}&years=${p.years}&growth_pct=${p.growth_pct}&reinvest=${p.reinvest}`),
  wealthIpoSim: (p: { offer_price: number; alloc_shares: number; subscribe_amount: number }) =>
    request<IpoSim>(`/api/data/wealth/ipo-sim?offer_price=${p.offer_price}&alloc_shares=${p.alloc_shares}&subscribe_amount=${p.subscribe_amount}`),
  wealthDividendPicks: (top = 12) => request<DividendPicks>(`/api/data/wealth/dividend-picks?top=${top}`),
  wealthIpoSchedule: () => request<IpoSchedule>(`/api/data/wealth/ipo-schedule`),
};
